webapp/components/NotificationMenu/NotificationMenu.vue

```diff
@@ -41,54 +46,39 @@ export default {
     async markAsRead(notificationSourceId) {
       const variables = { id: notificationSourceId }
       try {
-        const {
-          data: { markAsRead },
-        } = await this.$apollo.mutate({
+        await this.$apollo.mutate({
           mutation: markAsReadMutation(this.$i18n),
           variables,
         })
-        if (!(markAsRead && markAsRead.read === true)) return
-        this.displayedNotifications = this.displayedNotifications.map(n => {
-          return this.equalNotification(n, markAsRead) ? markAsRead : n
-        })
       } catch (err) {
         this.$toast.error(err.message)
       }
     },
-    equalNotification(a, b) {
-      return a.from.id === b.from.id && a.createdAt === b.createdAt && a.reason === b.reason
-    },
   },
   computed: {
-    notificationsCount() {
-      return (this.displayedNotifications || []).length
-    },
     unreadNotificationsCount() {
-      let countUnread = 0
-      if (this.displayedNotifications) {
-        this.displayedNotifications.forEach(notification => {
-          if (!notification.read) countUnread++
-        })
-      }
-      return countUnread
+      const result = this.notifications.reduce((count, notification) => {
+        return notification.read ? count : count + 1
+      }, 0)
+      return result
     },
   },
   apollo: {
     notifications: {
       query() {
         return notificationQuery(this.$i18n)
       },
+      variables() {
+        return {
+          read: false,
+          orderBy: 'updatedAt_desc',
+        }
+      },
       pollInterval: NOTIFICATIONS_POLL_INTERVAL,
-      update(data) {
-        const newNotifications = data.notifications.filter(newN => {
-          return !this.displayedNotifications.find(oldN => this.equalNotification(newN, oldN))
-        })
-        this.displayedNotifications = newNotifications
-          .concat(this.displayedNotifications)
-          .sort((a, b) => {
-            return new Date(b.createdAt) - new Date(a.createdAt)
-          })
-        return data.notifications
+      update({ notifications }) {
+        return unionBy(notifications, this.notifications, notification => notification.id).sort(
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-11-08T13:43:26Z" title="Friday, November 8th 2019, 2:43:26 pm +01:00">Nov 8, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -41,54 +46,39 @@ export default {
     async markAsRead(notificationSourceId) {
       const variables = { id: notificationSourceId }
       try {
-        const {
-          data: { markAsRead },
-        } = await this.$apollo.mutate({
+        await this.$apollo.mutate({
           mutation: markAsReadMutation(this.$i18n),
           variables,
         })
-        if (!(markAsRead && markAsRead.read === true)) return
-        this.displayedNotifications = this.displayedNotifications.map(n => {
-          return this.equalNotification(n, markAsRead) ? markAsRead : n
-        })
       } catch (err) {
         this.$toast.error(err.message)
       }
     },
-    equalNotification(a, b) {
-      return a.from.id === b.from.id && a.createdAt === b.createdAt && a.reason === b.reason
-    },
   },
   computed: {
-    notificationsCount() {
-      return (this.displayedNotifications || []).length
-    },
     unreadNotificationsCount() {
-      let countUnread = 0
-      if (this.displayedNotifications) {
-        this.displayedNotifications.forEach(notification => {
-          if (!notification.read) countUnread++
-        })
-      }
-      return countUnread
+      const result = this.notifications.reduce((count, notification) => {
+        return notification.read ? count : count + 1
+      }, 0)
+      return result
     },
   },
   apollo: {
     notifications: {
       query() {
         return notificationQuery(this.$i18n)
       },
+      variables() {
+        return {
+          read: false,
+          orderBy: 'updatedAt_desc',
+        }
+      },
       pollInterval: NOTIFICATIONS_POLL_INTERVAL,
-      update(data) {
-        const newNotifications = data.notifications.filter(newN => {
-          return !this.displayedNotifications.find(oldN => this.equalNotification(newN, oldN))
-        })
-        this.displayedNotifications = newNotifications
-          .concat(this.displayedNotifications)
-          .sort((a, b) => {
-            return new Date(b.createdAt) - new Date(a.createdAt)
-          })
-        return data.notifications
+      update({ notifications }) {
+        return unionBy(notifications, this.notifications, notification => notification.id).sort(
```


Wow. This is even shorter !!! 😍

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-11-08T16:52:52Z" title="Friday, November 8th 2019, 5:52:52 pm +01:00">Nov 8, 2019</time>_

---


thanks to @roschaefer ;)

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-11-08T20:00:17Z" title="Friday, November 8th 2019, 9:00:17 pm +01:00">Nov 8, 2019</time>_

---


👍🏼